### PR TITLE
Refactor validation tests

### PR DESCRIPTION
**Pull request by elray1**
In #41, we refactored some validations that were duplicated in `transform_point_model_out` and `transform_quantile_model_out` into a new/separate function, `validate_model_out_target_obs` that is called from each of those transform functions.

The main purpose of this PR is to tidy up the unit tests in an analogous way, so that tests of the transform functions that were really about those validation checks how just test the validation function instead.  This reduces some duplication in tests.

Along the way I found a couple of errors in test setup which I have also fixed, and also added another check to `validate_model_out_target_obs`.

### CODE REVIEW

File: tests/testthat/test-transform_point_model_out.R [hunk 1]
```diff
@@ -1,26 +1,4 @@
...
-    ))
-  )
-  expect_error(
-    suppressMessages(transform_point_model_out(
-      model_out_tbl = data.frame(),
-      target_observations = data.frame(),
-      output_type = "mean"
-    ))
-  )
-
```
Comment (elray1) [inline tests/testthat/test-transform_point_model_out.R:23]: moved to test-validate_model_out_target_obs.R

File: tests/testthat/test-transform_point_model_out.R [hunk 2]
```diff
@@ -60,26 +38,8 @@ test_that("model_out_tbl_1 output is valid", {
   exp_forecast <- utils::read.csv(
     test_path("testdata/exp_forecast_1.csv")
   )
-  class(exp_forecast) <- c("forecast", "forecast_point", "data.table", "data.frame")
-  expect_equal(act_forecast, exp_forecast, ignore_attr = "class")
-})
-
-test_that("test target_observations has observation column", {
```
Comment (elray1) [inline tests/testthat/test-transform_point_model_out.R:67]: moved to moved to test-validate_model_out_target_obs.R

File: tests/testthat/test-transform_point_model_out.R [hunk 3]
```diff
@@ -89,30 +49,22 @@ test_that("model_out_tbl_1 columns are valid", {
   model_out_tbl_1 <- utils::read.csv(test_path("testdata/model_out_tbl_point_1.csv")) |>
     dplyr::rename(loc = location, trgt = target, date = target_end_date)
 
-  target_observations_1 <- utils::read.csv(test_path("testdata/target_data_1.csv"))
+  target_observations_1 <- utils::read.csv(test_path("testdata/target_data_1.csv")) |>
+    dplyr::rename(loc = location, date = target_end_date)
```
Comment (elray1) [inline tests/testthat/test-transform_point_model_out.R:53]: we should have done this renaming previously to test for a valid merge.  the fact that this test passed previously was a bug (which is now fixed, i.e., the test without this renaming now fails).

File: tests/testthat/test-transform_point_model_out.R [hunk 4]
```diff
@@ -89,30 +49,22 @@ test_that("model_out_tbl_1 columns are valid", {
...
     output_type = "mean"
   )
   exp_forecast <- utils::read.csv(test_path("testdata/exp_forecast_1.csv")) |>
     dplyr::rename(loc = location, trgt = target, date = target_end_date)
-  class(exp_forecast) <- c("forecast", "forecast_point", "data.table", "data.frame")
+  class(exp_forecast) <- c("forecast_point", "forecast", "data.table", "data.frame")
   expect_equal(act_forecast, exp_forecast, ignore_attr = "class")
-
-  # Error when missing any of: model_id, output_type, output_type_id, value
-  expect_error(
```
Comment (elray1) [inline tests/testthat/test-transform_point_model_out.R:104]: this is a duplicate of another test, so it is just removed.

File: tests/testthat/test-transform_point_model_out.R [hunk 5]
```diff
@@ -129,24 +81,6 @@ test_that("model_out_tbl_1 has any rows", {
   )
 })
 
-test_that("model_out_tbl columns match target_observations columns", {
```
Comment (elray1) [inline tests/testthat/test-transform_point_model_out.R:132]: moved to moved to test-validate_model_out_target_obs.R and updated

File: tests/testthat/test-transform_quantile_model_out.R [hunk 6]
```diff
@@ -1,18 +1,3 @@
-test_that("inputs are valid", {
```
Comment (elray1) [inline tests/testthat/test-transform_quantile_model_out.R:1]: moved to moved to test-validate_model_out_target_obs.R

File: tests/testthat/test-transform_quantile_model_out.R [hunk 7]
```diff
@@ -22,25 +7,8 @@ test_that("model_out_tbl_1 output is valid", {
   )
 
   exp_forecast <- utils::read.csv(test_path("testdata/exp_forecast_2.csv"))
-  class(exp_forecast) <- c("forecast", "forecast_quantile", "data.table", "data.frame")
-  expect_equal(act_forecast, exp_forecast, ignore_attr = "class")
-})
-
-test_that("test target_observations has observation column", {
```
Comment (elray1) [inline tests/testthat/test-transform_quantile_model_out.R:29]: moved to moved to test-validate_model_out_target_obs.R

File: tests/testthat/test-transform_quantile_model_out.R [hunk 8]
```diff
@@ -61,21 +30,12 @@ test_that("model_out_tbl_1 columns are valid", {
     test_path("testdata/exp_forecast_2.csv")
   ) |>
     dplyr::rename(loc = location, trgt = target, date = target_end_date)
-  class(exp_forecast) <- c("forecast", "forecast_quantile", "data.table", "data.frame")
+  class(exp_forecast) <- c("forecast_quantile", "forecast", "data.table", "data.frame")
   expect_equal(act_forecast, exp_forecast, ignore_attr = "class")
-
-  # Error when missing any of: model_id, output_type, output_type_id, value
-  expect_error(
```
Comment (elray1) [inline tests/testthat/test-transform_quantile_model_out.R:68]: duplicates another test, so removed

File: tests/testthat/test-transform_quantile_model_out.R [hunk 9]
```diff
@@ -91,23 +51,6 @@ test_that("model_out_tbl_1 has any rows", {
   )
 })
 
-test_that("model_out_tbl columns match target_observations columns", {
```
Comment (elray1) [inline tests/testthat/test-transform_quantile_model_out.R:94]: moved to moved to test-validate_model_out_target_obs.R

File: tests/testthat/test-transform_point_model_out.R [hunk 10]
```diff
@@ -60,26 +38,8 @@ test_that("model_out_tbl_1 output is valid", {
   exp_forecast <- utils::read.csv(
     test_path("testdata/exp_forecast_1.csv")
   )
-  class(exp_forecast) <- c("forecast", "forecast_point", "data.table", "data.frame")
-  expect_equal(act_forecast, exp_forecast, ignore_attr = "class")
```
[the remark this answers is not in the document]
Comment (zkamvar) [reply tests/testthat/test-transform_point_model_out.R:64]: This makes sense. I was confused as to why the test was failing on CI and not my machine and chalked it down to a non-deterministic assignment of classes, but it turned out that scoringutils was outdated on my machine. 

The fix from scoringutils was in https://github.com/epiforecasts/scoringutils/pull/871

**Review by zkamvar — CHANGES_REQUESTED** [review]
This looks good. I made my notes below and a couple of broad observations:

0. I would turn the warning into an error. Warnings are a weird sort of alert in R. Messages inform the user of processes that are going on and Errors tell the user that something is wrong. Warnings are a way of telling the user that something is weird and maybe wrong, but it shows up _after_ the damage has been done. 
1. It's not a bad idea to test for the same error/warning in multiple places
2. just a small note for testing errors: if it comes from another package, we may run into issues in the future if error messages change.

File: tests/testthat/test-validate_model_out_target_obs.R [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+  cli::test_that_cli("alert", {
+    expect_snapshot(
+      val_result <- validate_model_out_target_obs(
+        model_out_tbl = model_out_tbl_1 |>
+          dplyr::rename(loc = location),
+        target_observations = target_observations_1
+      )
+    )
+  })
+})
```
[the remark this answers is not in the document]
Comment (elray1) [reply tests/testthat/test-validate_model_out_target_obs.R:94]: thanks!  I struggled to understand how this was supposed to work from their docs 😅

File: tests/testthat/test-transform_point_model_out.R [hunk 12]
```diff
@@ -43,24 +43,22 @@ test_that("model_out_tbl_1 output is valid", {
...
-      regexp = "unexpected column"
+
+  expect_snapshot(
+    transform_point_model_out(
+      model_out_tbl = model_out_tbl_1,
+      target_observations = target_observations_1,
+      output_type = "mean"
     )
-  })
+  )
```
Comment (zkamvar) [inline tests/testthat/test-transform_point_model_out.R:61]: Oh I see now why this test was failing! `cli::cli_alert_warning()` produces a stylized `message` and `cli::cli_warn()` produces an actual warning... I had gotten them confused.

**Review by zkamvar — APPROVED** [review]
Thank you for the changes. There is still linting, but I guarantee that's my fault (I litter whitespace at the end of lines)

File: R/validate.R [hunk 13]
```diff
@@ -42,6 +42,17 @@ validate_model_out_target_obs <- function(model_out_tbl, target_observations) {
...
+  t_o_cols <- colnames(target_observations)
+  expected_cols_superset <- c(task_id_cols, "output_type", "output_type_id", "observation")
+  unexpected_cols <- t_o_cols[!t_o_cols %in% expected_cols_superset]
+  if (length(unexpected_cols) > 0) {
+    cli::cli_alert_warning(
+      c(
+        "`target_observations` had {length(unexpected_cols)} unexpected column{?s} {.val {unexpected_cols}};",
+        " expected the columns of `target_observations` to be a subset of {.val {expected_cols_superset}}."
+      )
+    )
```
[the remark this answers is not in the document]
Comment (elray1) [reply R/validate.R:54]: Good point.  I was thinking there might be reasons someone might have other columns in there, but now I can't think of any (good) reasons to do that.  I'll update this to an error.

File: tests/testthat/test-transform_point_model_out.R (hunk 12 above)
Comment (elray1) [reply tests/testthat/test-transform_point_model_out.R]: i hadn't actually seen `cli_warn` -- good to know about that one.  regardless, this is now an error...

File: R/validate.R (hunk 13 above)
[the remark this answers is not in the document]
Comment (elray1) [reply R/validate.R]: in a new commit [1d4a0bb](https://github.com/hubverse-org/hubEvals/pull/47/commits/1d4a0bb598c227f6d79d8f5767044d8307dab4c2), this is now an error

**Comment by elray1** [comment]
Thanks for the constructive review!  I'll merge this now.

**Pull request merged**